<a href="https://colab.research.google.com/github/prithvikolla/GitHub_Repository_Analysis/blob/main/GRA_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

import io

import requests

urlfile="https://raw.githubusercontent.com/prithvikolla/GitHub_Repository_Analysis/main/2015-01-01-0.csv"

url=requests.get(urlfile).content

df_GRA =pd.read_csv(io.StringIO(url.decode('utf-8')))


In [2]:
df_GRA.head()

id  ...                                  org.avatar_url
0  2489368070  ...                                              ""
1  2489368072  ...                                              ""
2  2489368089  ...                                              ""
3  2489368095  ...  https://avatars.githubusercontent.com/u/82592?
4  2489368104  ...                                              ""

[5 rows x 182 columns]

In [3]:
df_GRA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7702 entries, 0 to 7701
Columns: 182 entries, id to org.avatar_url
dtypes: bool(1), float64(1), int64(3), object(177)
memory usage: 10.6+ MB


In [4]:
unique_actors = list (df_GRA['actor.login'].unique())

In [5]:
len(unique_actors)

3648

In [6]:
len(df_GRA['actor.id'].unique())

3648

1. How many unique contributors for each group?

- Contributors can be inferred as , users who have done push event and finding the distinct users by "actor.id" attribute and "type" attribute.

In [7]:
id = [ ]
for i in range(len(df_GRA['type'])): 
  if (df_GRA['type'][i] == 'PushEvent'):
    id.append(df_GRA['actor.id'][i])

In [8]:
def uniq(input):
  output = [ ]
  for x in input:
    if x not in output:
      output.append(x)
  return output
id_unique = uniq(input = id)

In [9]:
len(id_unique)

2327

In [10]:
print(len(id) == len(id_unique))

False


hence, all the push events done have same user contributing multiple times

- so, number of unique contributions are 2327 by the analysis done till now.

2. What's the distribution of the number of contributors among the all projects here? 
 - what is the mean?
 - what is the standard deviation? 
 - Is it a normal distribution?

In [11]:
from nltk import Counter

In [12]:
event_Count = Counter(df_GRA['type'])

In [13]:
event_Count.most_common()

[('PushEvent', 4280),
 ('CreateEvent', 815),
 ('IssueCommentEvent', 650),
 ('WatchEvent', 642),
 ('IssuesEvent', 373),
 ('PullRequestEvent', 315),
 ('ForkEvent', 213),
 ('DeleteEvent', 141),
 ('GollumEvent', 90),
 ('PullRequestReviewCommentEvent', 85),
 ('CommitCommentEvent', 56),
 ('ReleaseEvent', 24),
 ('MemberEvent', 16),
 ('PublicEvent', 2)]

As we can see,In the top 3 common events done in Github. We find that users did largely "Push Event" and the second one is "Create Event"

In [14]:
#Check wether we have entire columns in the whole data set with null data.
null_data = pd.DataFrame()
for i in df_GRA.columns:
  if (len(df_GRA[i])== 0):
    null_data.insert(0,i,df_GRA[i])

In [15]:
len(null_data)
## No we do not have entire data columns in the whole data set with null values.

0

In [23]:
user_ID = [ ]
repo_ID = [ ]
for i in range(len(df_GRA)):
 if (df_GRA['type'][i] == 'PushEvent'):
   user_ID.append(df_GRA['actor.id'][i])
   repo_ID.append(df_GRA['repo.id'][i])
import matplotlib.pyplot as plt
from nltk import Counter
count_ID = Counter(user_ID)
top_10_ID  = count_ID.most_common(10)

In [25]:
df2

actor.id
7                       [538746]
137         [28648528, 16931265]
141                   [14843113]
281                    [8039659]
375         [28546246, 28574862]
                    ...         
10361421              [28678011]
10361430              [28678127]
10361441              [20042152]
10361442               [1334170]
10361459              [28527093]
Length: 3648, dtype: object

In [28]:
top_10_ID[:3]

[(280212, 79), (2033581, 42), (367618, 35)]

In [32]:
dicti_ry = {k: v for k, v in top_10_ID}

In [39]:
list_Top10_User = pd.DataFrame({'top10_userID': list (dicti_ry.keys()),'Num_PusEve' :  list (dicti_ry.values())})


The above result shows the top 10 users who contributed for all the projects

In [40]:
list_Top10_User

top10_userID  Num_PusEve
0        280212          79
1       2033581          42
2        367618          35
3       9201970          29
4       6158630          24
5       9452791          24
6       1684950          20
7       1181205          18
8       6214365          18
9         50891          15

In [44]:

df_Push = df_GRA[df_GRA['type'] == 'PushEvent']

In [70]:
group = df_Push.groupby('actor.id')
df_u = group.apply(lambda x: x['repo.id'].unique())

In [72]:
my_li = list (set(df_Push['actor.id']))

In [91]:
df_u

actor.id
7                       [538746]
137                   [28648528]
375         [28546246, 28574862]
1106                   [2393766]
1377                  [28646975]
                    ...         
10361273              [28677442]
10361298              [28677764]
10361312              [28677674]
10361315              [28677587]
10361338              [28677651]
Length: 2327, dtype: object

In [ ]:
val = [x for x in df_u]

In [92]:
li = pd.DataFrame({'user_push': sorted(my_li),'rep_user':val})

In [93]:
li.head()

user_push              rep_user
0          7              [538746]
1        137            [28648528]
2        375  [28546246, 28574862]
3       1106             [2393766]
4       1377            [28646975]

In [101]:
arr = [ ]
for i in range(len(list_Top10_User)):
  for j in range(len(li)):
    if (list_Top10_User['top10_userID'][i] == li['user_push'][j]):
       arr.append(li['rep_user'][j])


In [106]:
arr

[array([21481110]),
 array([6058151, 6058123, 6058234, 6031942]),
 array([  756568,   860274, 14148013, 11673915,   747311,   779156,
         6485371, 23811389,   756841,  2387146,  2387049,  5054494,
        10040198, 10038467,  6756500, 10039899,  8694598, 11672686,
        26532168, 16819408,  6756503]),
 array([25173910]),
 array([18126008, 15100395]),
 array([28676949, 28677711, 25941039, 26978339, 25940223]),
 array([28650038]),
 array([28657751]),
 array([28669435]),
 array([28677563, 23131813, 26407499])]

In [109]:
list_Top10_User['rep_pushOcc'] = arr
list_Top10_User['rep_Count'] = list_Top10_User['rep_pushOcc'].apply(lambda x: len(x))

In [110]:
list_Top10_User

top10_userID  ...  rep_Count
0        280212  ...          1
1       2033581  ...          4
2        367618  ...         21
3       9201970  ...          1
4       6158630  ...          2
5       9452791  ...          5
6       1684950  ...          1
7       1181205  ...          1
8       6214365  ...          1
9         50891  ...          3

[10 rows x 4 columns]

From the above table I wanted to show the top 10 contributors and how many projects they were working on.

we see that,

1. user: 367618 was working parallely in more than 20 projects.
2. user: 28021 worked only on one project and even though push events performed are highest, we cannot conclude that he is the highest contributor among the 10 without considering other factors like length of code contributed per push event.